server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1053,31 +1233,33 @@ public static Map<String, Set<String>> resolveSearchRoutingAllIndices(Metadata m
      * @return true if the provided array maps to all indices, false otherwise
      */
     public static boolean isAllIndicesExpression(Collection<ResolvedExpression> aliasesOrIndices) {
-        return aliasesOrIndices == null || aliasesOrIndices.isEmpty() || isExplicitAllPatternExpression(aliasesOrIndices);
+        return isAllIndices(aliasesOrIndices, ResolvedExpression::resource);
     }
 
     /**
-     * Identifies whether the array containing index names given as argument explicitly refers to all indices
-     * The empty or null array doesn't explicitly map to all indices
+     * Identifies whether the array containing index names given as argument refers to all indices
+     * The empty or null array identifies all indices
      *
      * @param aliasesOrIndices the array containing index names
-     * @return true if the provided array explicitly maps to all indices, false otherwise
+     * @return true if the provided array maps to all indices, false otherwise
      */
-    static boolean isExplicitAllPatternExpression(Collection<ResolvedExpression> aliasesOrIndices) {
-        return aliasesOrIndices != null
-            && aliasesOrIndices.size() == 1
-            && Metadata.ALL.equals(aliasesOrIndices.iterator().next().resource());
+    public static boolean isAllIndices(Collection<String> aliasesOrIndices) {
+        return isAllIndices(aliasesOrIndices, Function.identity());
```

Comment:
I'm kinda indifferent about changing it to `String[]` over the `ArrayList` since it looks like the callers to this use `Arrays.asList` which just call the ArrayList constructor that takes the string array in. No list copies or anything like that have to happen, it just creates the wrapper object around the array. Not really sure if that would move the needle performance wise or not.